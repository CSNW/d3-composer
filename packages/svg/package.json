{
  "name": "@d3-composer/svg",
  "version": "0.3.1",
  "description": "svg components for d3-composer",
  "author": "Tim Hall <tim.hall.engr@gmail.com>",
  "license": "MIT",
  "repository": "https://github.com/CSNW/d3-composer.git",
  "contributors": [
    "CSNW"
  ],
  "bugs": {
    "url": "https://github.com/CSNW/d3-composer/issues"
  },
  "homepage": "https://github.com/CSNW/d3-composer",
  "keyword": [
    "d3",
    "d3-composer",
    "svg",
    "chart",
    "visualization"
  ],
  "main": "dist/d3-composer-svg.js",
  "unpkg": "dist/d3-composer-svg.min.js",
  "module": "src/index.js",
  "scripts": {
    "test": "exit 0",
    "build": "rollup -c"
  },
  "dependencies": {
    "@d3-composer/utils": "^0.3.1",
    "d3-axis": "^1",
    "d3-selection": "^1",
    "d3-shape": "^1"
  },
  "devDependencies": {
    "@d3-composer/grid": "^0.3.1"
  },
  "files": ["dist", "src", "!src/__tests__", "!src/__helpers__"]
}
