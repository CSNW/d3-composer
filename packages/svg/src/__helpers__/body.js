import { select } from 'd3';

export default function body() {
  let selection;
  let svg;
  const instances = [];

  return {
    setup() {
      selection = select(document.body);
      svg = selection.append('svg');
    },
    teardown() {
      for (const instance of instances) {
        instance.remove();
      }
      svg.remove();
    },

    append(type) {
      const instance = selection.append(type);
      instances.push(instance);

      return instance;
    },
    layer() {
      return svg.append('g');
    }
  };
}
