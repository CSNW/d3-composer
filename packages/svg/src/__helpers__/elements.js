const default_size = { width: 100, height: 100 };

export class SVGElement {
  constructor(options = {}) {
    const { size = default_size } = options;

    this.size = size;
  }

  getBBox() {
    return this.size;
  }
}
