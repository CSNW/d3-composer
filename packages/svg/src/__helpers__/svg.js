import { select } from 'd3';

export function svg() {
  let instance;

  function layer() {
    return instance.append('g');
  }

  layer.setup = () => {
    instance = select(document.body).append('svg');
  };
  layer.teardown = () => {
    instance.remove();
  };

  return layer;
}
