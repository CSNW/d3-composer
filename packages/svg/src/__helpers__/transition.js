export function end(transition) {
  return new Promise(resolve => {
    transition.on('end', () => resolve());
  });
}
