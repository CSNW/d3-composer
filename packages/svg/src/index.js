export { default as chart } from './chart';
export { default as layout } from './layout';
export { default as layer } from './layer';
export { default as series } from './series';
export { default as line } from './line';
export { default as bars } from './bars';
export { default as scatter } from './scatter';
export { default as area } from './area';
export { default as labels } from './labels';
export { axisLeft, axisRight, axisBottom, axisTop } from './axis';
export { default as text } from './text';
export { default as legend } from './legend';
export { default as gridlines } from './gridlines';
export { symbolLine } from './symbols';
export { size, interpolatePath, translateXY } from './utils';
