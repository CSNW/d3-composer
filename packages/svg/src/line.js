import { passthrough, array, toStyle } from '@d3-composer/utils';
import { line as d3_line } from 'd3-shape';
import { interpolatePath } from './utils';

export default function line(selection, props = {}) {
  let {
    data, // = passthrough,
    x,
    y,
    key,
    style,
    class: className,
    transition,
    curve,
    interpolate,
    defined
  } = props;
  data = data || passthrough; // TEMP (https://github.com/rollup/rollup/issues/2445)

  const line = d3_line()
    .x(x)
    .y(y);
  if (curve) line.curve(curve);
  if (defined) line.defined(defined);

  const path = selection.selectAll('path').data(array(data), key);

  path.exit().remove();

  path
    .enter()
    .append('path')
    .attr('d', line)
    .merge(path)
    .attr('fill', 'none')
    .attr('style', toStyle(style))
    .attr('class', className)
    .transition(transition)
    .call(interpolatePath, line, interpolate);

  return selection;
}
