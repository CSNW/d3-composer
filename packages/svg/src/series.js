import { passthrough, toStyle, Series } from '@d3-composer/utils';
import { childNodes } from './utils';

const seriesKey = (d, i) => (d && d.key != null ? d.key : i);

export default function series(selection, props) {
  let {
    data, // = passthrough,
    key = seriesKey,
    style,
    class: className
  } = props;
  data = data || passthrough; // TEMP

  const layers = selection
    .selectAll(childNodes)
    .filter('[data-series]')
    .data(data, key);

  layers.exit().remove();

  return layers
    .enter()
    .append('g')
    .attr('data-series', key)
    .property(Series, d => d)
    .merge(layers)
    .attr('class', className)
    .attr('style', toStyle(style))
    .order();
}
