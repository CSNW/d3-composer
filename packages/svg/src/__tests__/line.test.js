import { scaleLinear, transition as d3_transition } from 'd3';
import { svg } from '../__helpers__/svg';
import { end } from '../__helpers__/transition';
import series from '../series';
import line from '../line';

const fixture = svg();

beforeAll(fixture.setup);
afterAll(fixture.teardown);

const xScale = scaleLinear()
  .domain([0, 100])
  .range([0, 100]);
const x = d => xScale(d.x);

const yScale = scaleLinear()
  .domain([0, 100])
  .range([100, 0]);
const y = d => yScale(d.y);

test('should render simple line', async () => {
  const transition = d3_transition().duration(0);
  const selection = line(fixture(), {
    data: [{ x: 0, y: 0 }, { x: 100, y: 100 }],
    x,
    y,
    transition
  });

  await end(transition);

  expect(selection.node()).toMatchSnapshot();
});

test('should use style and class', async () => {
  const transition = d3_transition().duration(0);
  const selection = line(
    series(fixture(), {
      data: [
        { class: 'a', values: [{ x: 0, y: 0 }, { x: 100, y: 100 }] },
        { class: 'b', values: [{ x: 0, y: 100 }, { x: 100, y: 0 }] }
      ],
      class: series => series.class
    }),
    {
      data: series => series.values,
      x,
      y,
      transition,

      style: {
        stroke: (_, i) => (i % 2 === 0 ? 'blue' : 'red')
      }
    }
  );

  await end(transition);

  expect(selection.node()).toMatchSnapshot();
});
