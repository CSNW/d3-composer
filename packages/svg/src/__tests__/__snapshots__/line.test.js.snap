// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render simple line 1`] = `
<g>
  <path
    d="M0,100L100,0"
    fill="none"
    stroke="currentColor"
  />
</g>
`;

exports[`should use style and class 1`] = `
<g
  class="a"
  data-series="0"
>
  <path
    d="M0,100L100,0"
    fill="none"
    stroke="currentColor"
    style="stroke: blue;"
  />
</g>
`;
